pkg/graveler/joined_diff_iterator.go

```diff
@@ -0,0 +1,116 @@
+package graveler
+
+import "bytes"
+
+// JoinedDiffIterator calculate the union diff between 2 iterators.
+// The output iterator yields a single result for each key.
+// If a key exist in the 2 iterators, iteratorA value prevails.
+type JoinedDiffIterator struct {
+	iterA        DiffIterator
+	iterAHasMore bool
+	iterB        DiffIterator
+	iterBHasMore bool
+	p            DiffIterator
+	started      bool
+}
+
+func NewJoinedDiffIterator(iterA DiffIterator, iterB DiffIterator) *JoinedDiffIterator {
+	return &JoinedDiffIterator{
+		iterA:        iterA,
+		iterAHasMore: true,
+		iterB:        iterB,
+		iterBHasMore: true,
+		p:            nil,
+	}
+}
+
+func (c *JoinedDiffIterator) Next() bool {
+	valA := c.iterA.Value()
+	valB := c.iterB.Value()
+	switch {
+	case !c.started:
+		// first
+		c.iterAHasMore = c.iterA.Next()
+		c.iterBHasMore = c.iterB.Next()
+		c.started = true
+	case !c.iterAHasMore && !c.iterBHasMore:
+		// last
+		return false
+	case !c.iterAHasMore:
+		// iterA is done
+		c.p = c.iterB
+		c.iterBHasMore = c.iterB.Next()
+	case !c.iterBHasMore:
+		// iterB is done
+		c.p = c.iterA
+		c.iterAHasMore = c.iterA.Next()
+	case bytes.Equal(valA.Key, valB.Key):
+		c.iterAHasMore = c.iterA.Next()
+		c.iterBHasMore = c.iterB.Next()
+	case bytes.Compare(valA.Key, valB.Key) < 0:
```

Comment:
This traverses the keys _twice_.  Please compare once; it is fine to call a separate routine here, but that will need to be documented.  In fact you can return immediately in all the above cases, and then compute `bytes.Compare(val!.Key, valB.Key)` once.